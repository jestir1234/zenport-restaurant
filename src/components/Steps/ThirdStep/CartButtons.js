import React from "react";
import styled, { css, keyframes } from "react-emotion";

const fade = keyframes`
  from { opacity: 0; }
  50%  { opacity: .5 }
  to   { opacity: 1 }`;

const Icon = styled("div")(props => ({
  height: "20px",
  width: "20px",
  padding: "10px",
  margin: "10px",
  borderRadius: "50%",
  textAlign: "center",
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  background: "#d3f4e6",
  fontSize: "14px",
  "&:hover": {
    background: "#bcf9df"
  },
  animation: `${fade} 1s ease-in-out`,
  cursor: "pointer"
}));

const Container = styled("div")(props => ({
  display: props.selectedValue ? "flex" : "none"
}));

const CartButtons = props => {
  return (
    <Container selectedValue={props.selectedValue}>
      <Icon onClick={props.addOrder}>+</Icon>
      <Icon onClick={props.subtractOrder}>&mdash;</Icon>
    </Container>
  );
};

export default CartButtons;
