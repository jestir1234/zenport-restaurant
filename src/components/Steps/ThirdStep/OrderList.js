import React from "react";
import styled, { css } from "react-emotion";

const ListContainer = styled("div")(props => ({
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  alignItems: "flex-start"
}));

const ListItem = styled("div")(props => ({
  display: "flex",
  justifyContent: "space-evenly",
  marginTop: "5px"
}));

const Container = styled("div")(props => ({
  display: "flex",
  flexDirection: "column",
  width: "100%"
}));

const OrderList = props => {
  return (
    <div>
      <h3>Orders</h3>
      <ListContainer>
        {Object.keys(props.currentOrder).map((dish, idx) => {
          return (
            <ListItem key={idx}>
              {dish} {props.currentOrder[dish]}
            </ListItem>
          );
        })}
      </ListContainer>
    </div>
  );
};

export default OrderList;
