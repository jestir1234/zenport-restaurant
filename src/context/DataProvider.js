import React, { Component } from "react";
import data from "../data.json";

export const DataContext = React.createContext();

const parseRestaurants = () => {
  let restaurants = {};
  let uniqRestaurants = Array.from(
    new Set(data[0].dishes.map(dish => dish.restaurant))
  );
  uniqRestaurants.forEach(name => {
    restaurants[name] = { availableMeals: "", menu: [] };
  });

  data[0].dishes.forEach(dish => {
    let currentMenu = restaurants[dish.restaurant].menu;
    currentMenu.push(dish.name);
    restaurants[dish.restaurant] = {
      availableMeals: dish.availableMeals,
      menu: currentMenu
    };
  });
  return restaurants;
};

const parseMealTypes = restaurants => {
  let mealTypes = {};
  Object.keys(restaurants).forEach(restaurant => {
    let restaurantMealTypes = restaurants[restaurant]["availableMeals"];
    restaurantMealTypes.forEach(type => {
      if (mealTypes[type]) {
        mealTypes[type]["restaurants"][restaurant] = restaurants[restaurant];
      } else {
        mealTypes[type] = { restaurants: {} };
      }
    });
  });
  return mealTypes;
};

const defaultStepOptions = () => ({
  step1: {
    selectedMealType: null,
    selectedPartyCount: { value: 1, label: 1 }
  },
  step2: {
    selectedRestaurant: null
  },
  step3: {
    selectedMeal: null,
    selectedServings: { value: 1, label: 1 },
    currentOrder: {},
    validMealCount: false,
    errorMessage: ""
  }
});

class DataProvider extends Component {
  state = {
    mealTypes: parseMealTypes(parseRestaurants()),
    stepOptions: defaultStepOptions()
  };

  handleChange = (selectedOption, option, currentStep) => {
    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        [currentStep]: {
          ...prevState.stepOptions[currentStep],
          [option]: selectedOption
        }
      }
    }));
  };

  addOrder = order => {
    let newOrder = { ...this.state.stepOptions.step3.currentOrder, ...order };
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder: newOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  subtractOrder = dish => {
    let currentOrder = { ...this.state.stepOptions.step3.currentOrder };
    delete currentOrder[dish];
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  validateMealCount = () => {
    let currentOrders = this.state.stepOptions.step3.currentOrder;
    let totalMeals =
      Object.keys(currentOrders).length > 0
        ? Object.values(this.state.stepOptions.step3.currentOrder).reduce(
            (acc, val) => acc + val
          )
        : 0;

    let selectedPartyCount = this.state.stepOptions.step1.selectedPartyCount
      .value;
    let errorMessage = "";

    if (totalMeals > 10) {
      errorMessage = "Exceeded max total servings of 10.";
    }

    if (totalMeals < selectedPartyCount) {
      errorMessage =
        "Total servings must be equal to or greater than the no. of people in your group.";
    }

    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        step3: {
          ...prevState.stepOptions.step3,
          validMealCount: totalMeals <= 10 && totalMeals >= selectedPartyCount,
          errorMessage
        }
      }
    }));
  };

  render() {
    return (
      <DataContext.Provider
        value={{
          state: this.state,
          handleChange: (selectedOption, option, currentStep) =>
            this.handleChange(selectedOption, option, currentStep),
          addOrder: order => this.addOrder(order),
          subtractOrder: dish => this.subtractOrder(dish)
        }}
      >
        {this.props.children}
      </DataContext.Provider>
    );
  }
}

export default DataProvider;
