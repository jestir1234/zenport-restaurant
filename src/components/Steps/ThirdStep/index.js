import React from "react";
import styled from "react-emotion";
import { DataContext } from "../../../context/DataProvider";
import Selector from "../Selector";
import CartButtons from "./CartButtons";
import OrderList from "./OrderList";

// STYLES
const Container = styled("div")(props => ({
  display: "flex",
  flexDirection: "column",
  alignItems: "flex-start"
}));

const SelectorContainer = styled("div")(props => ({
  display: "flex",
  justifyContent: "space-between",
  width: "100%"
}));

const SelectorStyles = {
  flex: 1
};

// CONSTANTS
const SERVING_MAX = 10;

// HELPER FUNCTIONS
const onlyUnique = (value, index, self) => self.indexOf(value) === index;

class ThirdStep extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      step: "step3"
    };
  }

  render() {
    return (
      <DataContext.Consumer>
        {context => {
          const stepOptions = context.state.stepOptions;
          let selectedType;
          let restaurants;
          if (context.state.stepOptions.step1.selectedMealType) {
            selectedType =
              context.state.stepOptions.step1.selectedMealType.value;
            restaurants = context.state.mealTypes[selectedType]["restaurants"];
          }
          let selectedRestaurant = stepOptions.step2["selectedRestaurant"];
          let menus = [];
          if (selectedRestaurant) {
            menus =
              restaurants[stepOptions.step2["selectedRestaurant"]["value"]][
                "menu"
              ];
          }

          if (context.state.mealTypes[selectedType]) {
            restaurants = context.state.mealTypes[selectedType]["restaurants"];
          }
          const selectedServingValue =
            stepOptions[this.state.step]["selectedServings"];
          const selectedMealValue =
            stepOptions[this.state.step]["selectedMeal"];

          let mealName;
          let addedOrder = {};
          if (selectedMealValue) {
            mealName = selectedMealValue.value;
          }
          if (selectedServingValue) {
            addedOrder[mealName] = selectedServingValue.value;
          }

          return (
            <Container>
              <SelectorContainer>
                <Selector
                  styles={SelectorStyles}
                  title={"Please Select a Meal"}
                  selectedValue={selectedMealValue}
                  handleChange={e =>
                    context.handleChange(e, "selectedMeal", this.state.step)
                  }
                  options={menus.filter(onlyUnique).map(name => {
                    return {
                      value: name,
                      label: name
                    };
                  })}
                />

                <Selector
                  errorMessage={stepOptions[this.state.step]["errorMessage"]}
                  styles={SelectorStyles}
                  title={"Please enter no. of servings"}
                  selectedValue={selectedServingValue}
                  handleChange={e =>
                    context.handleChange(e, "selectedServings", this.state.step)
                  }
                  options={[...Array(SERVING_MAX)].map((count, idx) => {
                    return {
                      value: idx + 1,
                      label: idx + 1
                    };
                  })}
                />
              </SelectorContainer>
              <CartButtons
                addOrder={() => context.addOrder(addedOrder)}
                subtractOrder={() => context.subtractOrder(mealName)}
                selectedValue={selectedServingValue && selectedMealValue}
              />

              <OrderList
                currentOrder={
                  context.state.stepOptions[this.state.step]["currentOrder"]
                }
              />
            </Container>
          );
        }}
      </DataContext.Consumer>
    );
  }
}

export default ThirdStep;
