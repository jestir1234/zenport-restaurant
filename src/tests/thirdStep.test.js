import React from "react";
import ThirdStep from "../components/Steps/ThirdStep";
import renderer from "react-test-renderer";
import DataProvider from "../context/DataProvider";
import getElementWithContext from "react-test-context-provider";

const context = {
  handleChange: (e, selectedPartyCount, step) =>
    console.log("handleChangeFunction"),
  state: {
    stepOptions: {
      step1: {
        selectedPartyCount: 4
      },
      step2: {}
    },
    mealTypes: {
      lunch: {
        restaurants: {
          McDonalds: ["Burger", "Fries"]
        }
      }
    }
  }
};

var ThirdStepWithContext = getElementWithContext(
  context,
  <DataProvider>
    <ThirdStep />
  </DataProvider>
);

test("Renders the <SecondStep /> component", () => {
  const component = renderer.create(ThirdStepWithContext);
  let tree = component.toJSON();
  expect(tree);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
