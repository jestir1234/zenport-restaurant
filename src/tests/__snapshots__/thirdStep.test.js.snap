// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the <SecondStep /> component 1`] = `
<div
  className="css-1lxwves"
>
  <div
    className="css-z7mtfw"
  >
    <div
      className="css-1yftcma"
    >
      <h3
        className="css-v9loj0"
      >
        Please Select a Meal
      </h3>
      <div
        className="Select is-clearable is-searchable Select--single"
        style={undefined}
      >
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-2--value"
          >
            <div
              className="Select-placeholder"
            >
              Select...
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-2--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
      <span
        className="css-rwbn3g"
      />
    </div>
    <div
      className="css-1yftcma"
    >
      <h3
        className="css-v9loj0"
      >
        Please enter no. of servings
      </h3>
      <div
        className="Select has-value is-clearable is-searchable Select--single"
        style={undefined}
      >
        <input
          disabled={false}
          name="form-field-name"
          type="hidden"
          value="1"
        />
        <div
          className="Select-control"
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={undefined}
        >
          <span
            className="Select-multi-value-wrapper"
            id="react-select-3--value"
          >
            <div
              className="Select-value"
              style={undefined}
              title={undefined}
            >
              <span
                aria-selected="true"
                className="Select-value-label"
                id="react-select-3--value-item"
                role="option"
              >
                1
              </span>
            </div>
            <div
              className="Select-input"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-activedescendant="react-select-3--value"
                aria-describedby={undefined}
                aria-expanded="false"
                aria-haspopup="false"
                aria-label={undefined}
                aria-labelledby={undefined}
                aria-owns=""
                className={undefined}
                id={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                role="combobox"
                style={
                  Object {
                    "boxSizing": "content-box",
                    "width": "5px",
                  }
                }
                tabIndex={undefined}
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </span>
          <span
            aria-label="Clear value"
            className="Select-clear-zone"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            title="Clear value"
          >
            <span
              className="Select-clear"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&times;",
                }
              }
            />
          </span>
          <span
            className="Select-arrow-zone"
            onMouseDown={[Function]}
          >
            <span
              className="Select-arrow"
              onMouseDown={[Function]}
            />
          </span>
        </div>
      </div>
      <span
        className="css-rwbn3g"
      >
        
      </span>
    </div>
  </div>
  <div
    className="css-1hyfx7x"
  >
    <div
      className="css-1s2rvl5"
      onClick={[Function]}
    >
      +
    </div>
    <div
      className="css-1s2rvl5"
      onClick={[Function]}
    >
      â€”
    </div>
  </div>
  <div>
    <h3 />
    <div
      className="css-17mvcka"
    />
  </div>
</div>
`;
